File: crates/1bit-helm/src/session.rs
//! Session-wide configuration (server URL, auth, default model, system prompt).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SessionConfig {
    pub server_url: String,
    #[serde(default)]
    pub bearer: Option<String>,
    pub default_model: String,
    #[serde(default)]
    pub system_prompt: Option<String>,
}

impl SessionConfig {
    pub fn new(server_url: impl Into<String>, default_model: impl Into<String>) -> Self {
        Self {
            server_url: server_url.into(),
            bearer: None,
            default_model: default_model.into(),
            system_prompt: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn session_config_serde_roundtrip() {
        let cfg = SessionConfig {
            server_url: "http://127.0.0.1:8080".into(),
            bearer: Some("tok".into()),
            default_model: "bitnet-3b".into(),
            system_prompt: Some("You are Halo.".into()),
        };
        let json = serde_json::to_string(&cfg).unwrap();
        let back: SessionConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(cfg, back);
    }

    #[test]
    fn session_config_defaults_optional_fields() {
        let src = r#"{"server_url":"http://x","default_model":"m"}"#;
        let cfg: SessionConfig = serde_json::from_str(src).unwrap();
        assert!(cfg.bearer.is_none());
        assert!(cfg.system_prompt.is_none());
    }
}

File: cpp/helm/src/session.cpp
#include "onebit/helm/session.hpp"

namespace onebit::helm {

nlohmann::json to_json(const SessionConfig& c)
{
    nlohmann::json j = {
        {"server_url", c.server_url},
        {"default_model", c.default_model},
    };
    if (c.bearer)        j["bearer"]        = *c.bearer;
    if (c.system_prompt) j["system_prompt"] = *c.system_prompt;
    return j;
}

SessionConfig from_json(const nlohmann::json& j)
{
    SessionConfig c;
    c.server_url    = j.value("server_url",    std::string{});
    c.default_model = j.value("default_model", std::string{});
    if (j.contains("bearer") && j["bearer"].is_string()) {
        c.bearer = j["bearer"].get<std::string>();
    }
    if (j.contains("system_prompt") && j["system_prompt"].is_string()) {
        c.system_prompt = j["system_prompt"].get<std::string>();
    }
    return c;
}

} // namespace onebit::helm
